### PR TITLE
PICARD-1384: Create org.musicbrainz.Picard.appdata.xml

### PR DESCRIPTION
**Pull request by abhi-ohri**
Add an app description to MusicBrainz Picard. AppStream adds standardized descriptions of applications, that can be used by different application stores and installers.

# Summary
Add appdata xml file to the project.

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [X] Other
* **Describe this change in 1-2 sentences**:

# Problem
There is no standard metadata file available to help package distributors display a description of Picard.

* JIRA ticket (_optional_): [PICARD-1384](https://tickets.metabrainz.org/browse/PICARD-1384)

# Solution
Add Appdata  XML file

# Action

### CODE REVIEW

**Comment by samj1912** [comment]
I do not see any merit in just creating this file without any links to it in the installer. @phw any comments?

**Comment by phw** [comment]
Thanks, this is already looking good. Apart from the screenshots, which you already said you will add when ready, we need two changes to Picard to make this work as intended:

1. As mentioned in the task the appdata files needs to be installed to `/usr/share/metainfo`. You can achieve this by adding the following lines to `args['data_files']` in [setup.py#L656](https://github.com/metabrainz/picard/blob/master/setup.py#L656)
    ```
    ('share/metainfo', ['org.musicbrainz.Picard.appdata.xml']),
    ```

2. We should take the chance and rename the existing `picard.desktop` file to `org.musicbrainz.Picard.desktop`. For that you have to rename the file and also update the install line in [setup.py#L656](https://github.com/metabrainz/picard/blob/master/setup.py#L656)

The reason for the second change is that this file name is used in the appdata file as `<launchable type="desktop-id">`, and the naming with a reverse domain name like `org.musicbrainz.Picard` is the now recommended way of naming these files according to the [Freedesktop Desktop Entry Specification](https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s02.html). Also this is required by e.g. Flatpak.

**Comment by phw** [comment]
@abhi-ohri One minor thing: Please change the title of this pull request to be prefixed with `PICARD-1384: `. This will automatically link it to the ticket on JIRA.

**Comment by bochecha** [comment]
> We should take the chance and rename the existing picard.desktop file to org.musicbrainz.Picard.desktop. For that you have to rename the file and also update the install line in setup.py#L656

Indeed.

That's why I opened https://tickets.metabrainz.org/browse/PICARD-1393, before knowing @abhi-ohri would send his appdata file upstream, because I knew this would cause problems.

I recommend first renaming the resource files, then adding the appdata.

As it currently is, the appdata can't work.

**Comment by abhi-ohri** [comment]
@phw I have made the required changes.

File: setup.py [hunk 1]
```diff
@@ -654,14 +654,15 @@ def find_file_in_path(filename):
...
 
 args['data_files'] = [
+    ('share/metainfo', ['org.musicbrainz.Picard.appdata.xml']),
     ('share/icons/hicolor/16x16/apps', ['resources/images/16x16/picard.png']),
     ('share/icons/hicolor/24x24/apps', ['resources/images/24x24/picard.png']),
     ('share/icons/hicolor/32x32/apps', ['resources/images/32x32/picard.png']),
     ('share/icons/hicolor/48x48/apps', ['resources/images/48x48/picard.png']),
     ('share/icons/hicolor/128x128/apps', ['resources/images/128x128/picard.png']),
     ('share/icons/hicolor/256x256/apps', ['resources/images/256x256/picard.png']),
     ('share/icons/hicolor/scalable/apps', ['resources/img-src/picard.svg']),
```
Comment (bochecha) [inline setup.py]: All these icons also need to be renamed to `org.musicbrainz.Picard.*`.

Which means you must also modify the `Icon=` key in the desktop file. :wink:

Comment (abhi-ohri) [reply setup.py]: Should I also rename all icon png files in resources manually or they are auto-generated and I do not need to change their name. @bochecha

Comment (bochecha) [reply setup.py]: I see no indication that they are auto-generated, and they are even committed in Git.

Also, I just found that if you rename the image files, you also need to rename them there: https://github.com/metabrainz/picard/blob/master/resources/picard.qrc

To identify which one you need to rename, it's all the ones that get installed by `setup.py` in `${prefix}/share/icons/hicolor/${SIZE}/apps`.

File: org.musicbrainz.Picard.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    <content_attribute id="sex-appearance">none</content_attribute>
+    <content_attribute id="language-profanity">none</content_attribute>
+    <content_attribute id="language-humor">none</content_attribute>
+    <content_attribute id="language-discrimination">none</content_attribute>
+    <content_attribute id="social-chat">none</content_attribute>
+    <content_attribute id="social-info">none</content_attribute>
+    <content_attribute id="social-audio">none</content_attribute>
+    <content_attribute id="social-location">none</content_attribute>
+    <content_attribute id="social-contacts">none</content_attribute>
+    <content_attribute id="money-purchasing">none</content_attribute>
```
Comment (Freso) [inline org.musicbrainz.Picard.appdata.xml]: This seems inaccurate. Is Picard no longer asking people to donate to MetaBrainz?

File: org.musicbrainz.Picard.appdata.xml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  <launchable type="desktop-id">org.musicbrainz.Picard.desktop</launchable>
+
+  <screenshots>
+    <screenshot type="default">https://www.daitauha.fr/static/screenshots/picard/main.png</screenshot>
+  </screenshots>
+
+  <url type="homepage">https://picard.musicbrainz.org/</url>
+  <url type="bugtracker">https://tickets.metabrainz.org/projects/PICARD/issues/</url>
+  <url type="faq">https://picard.musicbrainz.org/docs/faq/</url>
+  <url type="help">https://community.metabrainz.org/c/picard</help>
```
Comment (Freso) [inline org.musicbrainz.Picard.appdata.xml]: https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-url suggests that the "help" type should be for "an online user's reference, a software manual or help page." I.e., documentation of some kind, not a support forum. Is it de facto used to link to forums (ie., did you see other AppStream files doing this) and the spec just doesn't reflect this (yet)?

File: org.musicbrainz.Picard.appdata.xml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    <content_attribute id="language-profanity">none</content_attribute>
+    <content_attribute id="language-humor">none</content_attribute>
+    <content_attribute id="language-discrimination">none</content_attribute>
+    <content_attribute id="social-chat">none</content_attribute>
+    <content_attribute id="social-info">none</content_attribute>
+    <content_attribute id="social-audio">none</content_attribute>
+    <content_attribute id="social-location">none</content_attribute>
+    <content_attribute id="social-contacts">none</content_attribute>
+    <content_attribute id="money-purchasing">none</content_attribute>
+    <content_attribute id="money-gambling">none</content_attribute>
```
Comment (Freso) [inline org.musicbrainz.Picard.appdata.xml]: AFAICT, [according to the spec](https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-content_rating), you only need to specify the ones that are *not* "none".

Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: I did not alter that because it was present in the base file and also in some appdata files I saw in other projects. Should I remove the unneeded ones?

File: org.musicbrainz.Picard.appdata.xml (hunk 3 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: I agree with you. Its a mistake I made. I am correcting it.
```suggestion
  <url type="help">https://picard.musicbrainz.org/docs/</url>
```

File: org.musicbrainz.Picard.appdata.xml (hunk 2 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: I checked with OARS asking for money donation is not same as money purchasing. Hence, I marked it as none.

File: setup.py (hunk 1 above)
Comment (abhi-ohri) [reply setup.py]: The images need to be refactored [here](https://github.com/metabrainz/picard/blob/5241c32c49e50aae49c2653f29713aea21ef6399/picard/ui/mainwindow.py) too.

**Comment by phw** [comment]
@abhi-ohri: After further checking the only thing breaking the builds right now is the rename of picard.ico. Since that one is used only for Windows it should not be renamed.

Since you already did this renaming here completely we can also keep it in if you fix just the .icon name. But I will add a separate ticket for this to track it in our changelog.

Thanks for the changes btw., good work, and sorry for the confusion. I just would have preferred to keep this change separate,but now that we already have it here let's not complicate things further :)

**Comment by abhi-ohri** [comment]
@phw Reverted the icon name and builds are passing. Any further instruction?

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (phw) [reply org.musicbrainz.Picard.appdata.xml]: Not sure really, is it ok to remove all of them if they are all none? @bochecha, since you added this initially any thoughts on it? I am not too familiar with OARS data inside appstream.

Comment (bochecha) [reply org.musicbrainz.Picard.appdata.xml]: If you remove all of them, then you end up not having any OARS metadata at all, which software centers interpret as "not rated" rather than "rated safe".

**Comment by bochecha** [comment]
> Since you already did this renaming here completely we can also keep it in if you fix just the .icon name. But I will add a separate ticket for this to track it in our changelog.

That's what I opened https://tickets.metabrainz.org/browse/PICARD-1393 for.

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (Freso) [reply org.musicbrainz.Picard.appdata.xml]: @bochecha That's contrary to the spec though. :/ It specifically says:
> In case the `<content_rating/>` tag is empty (no `<content_attribute/>` is present), it is assumed that the component was checked for age ratings and no age restrictions apply.

However, it that's the de facto standard, then that's how it is.

File: org.musicbrainz.Picard.appdata.xml (hunk 2 above)
Comment (Freso) [reply org.musicbrainz.Picard.appdata.xml]: Alright. The generator at https://hughsie.github.io/oars/generate.html has "Users are encouraged to donate real money, e.g. using Patreon" (which Picard *does* do) as `money-purchasing` = `mild`. But if you've asked OARS people and they say differently, then that's probably correct. :)

File: org.musicbrainz.Picard.appdata.xml [hunk 5]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+  <id>org.musicbrainz.Picard</id>
+  <project_license>GPL-2.0+</project_license>
```
Comment (Freso) [inline org.musicbrainz.Picard.appdata.xml]: @bochecha Hi :) https://spdx.org/licenses/GPL-2.0+ says it's deprecated. It doesn't say what it's deprecated in favour of, but https://spdx.org/licenses/ has "GPL v2.0 or later" as [`GPL-2.0-or-later`](https://spdx.org/licenses/GPL-2.0-or-later.html). Should the license be updated/changed here too?

File: org.musicbrainz.Picard.appdata.xml (hunk 2 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: @Freso Sorry for the error on my side. It indeed needs to be mild. It was a miscommunication on my part. I will correct it. I didn't know the donation column was in In-App Purchases.
```suggestion
    <content_attribute id="money-purchasing">mild</content_attribute>
```

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: @Freso @bochecha Since now we need to set content attribute money-purchasing to _mild_. I guess we could omit the unneeded ones.

**Comment by abhi-ohri** [comment]
@Freso @phw Should I link [PICARD-1393](https://tickets.metabrainz.org/browse/PICARD-1393) with the PR too?

File: org.musicbrainz.Picard.appdata.xml (hunk 2 above)
Comment (bochecha) [reply org.musicbrainz.Picard.appdata.xml]: Does the encouragement to donate appear inside the app?

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (bochecha) [reply org.musicbrainz.Picard.appdata.xml]: > In case the <content_rating/> tag is empty

Oh, maybe I'm mixing up between an empty `<content_rating/>` and no `<content_rating/>` at all?

I'll try an empty `<content_rating/>` in the Flathub build, and see what GNOME Software has to say about it. :slightly_smiling_face:

File: org.musicbrainz.Picard.appdata.xml (hunk 5 above)
Comment (bochecha) [reply org.musicbrainz.Picard.appdata.xml]: The problem is SPDX has multiple versions, and the appstream spec doesn't say which version it relies on.

However the examples use the deprecated form: https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-project_license

When in doubt, the best is to use `appstream-util validate-relax` to verify. Both seem to be accepted, so yes, let's move to the newer SPDX identifier. :slightly_smiling_face:

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: We can do that once it is resolved whether to use _money-purchasing_ to _none_ or _mild_ .

File: org.musicbrainz.Picard.appdata.xml (hunk 2 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: Yes it does on the About Page in the application.

**Comment by abhi-ohri** [comment]
I have added the screenshots as well.

File: org.musicbrainz.Picard.appdata.xml (hunk 2 above)
Comment (phw) [reply org.musicbrainz.Picard.appdata.xml]: We have a note in the about page and an actual "Donate..." menu entry, so I think setting this to mild would be just fair. Let's do this.

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (phw) [reply org.musicbrainz.Picard.appdata.xml]: Agreed with @abhi-ohri . So let's set money-purchasing to mild and remove everything else.

File: org.musicbrainz.Picard.appdata.xml (hunk 5 above)
Comment (Freso) [reply org.musicbrainz.Picard.appdata.xml]: @abhi-ohri :point_up_2: :)

Comment (phw) [reply org.musicbrainz.Picard.appdata.xml]: Let's got with the newer syntax "GPL-2.0-or-later" then

File: org.musicbrainz.Picard.appdata.xml (hunk 4 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]: SInce the `content_rating` now has a content-attribute `money-purchasing` we can remove the unneeded tags. @Freso What's your say?

File: org.musicbrainz.Picard.appdata.xml (hunk 5 above)
Comment (abhi-ohri) [reply org.musicbrainz.Picard.appdata.xml]:
```suggestion
  <project_license>GPL-2.0-or-later</project_license>
```
Done.

File: picard/ui/mainwindow.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
+﻿# -*- coding: utf-8 -*-
```
Comment (Freso) [inline picard/ui/mainwindow.py]: Did your editor add a [BOM](https://en.wikipedia.org/wiki/Byte_order_mark "byte order mark")? Or what changed on this line? BOMs are generally discouraged, and I don't see anything else that would have changed on this line.

Comment (phw) [reply picard/ui/mainwindow.py]: Good catch, indeed. I checked the file, there is a BOM. That should be removed for UTF-8 files.

Comment (abhi-ohri) [reply picard/ui/mainwindow.py]:
```suggestion
```

File: picard/ui/mainwindow.py [hunk 7]
```diff
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (zas) [inline picard/ui/mainwindow.py]: Why is this line removed ?

Comment (abhi-ohri) [reply picard/ui/mainwindow.py]: @zas It is a BOM(Byte Order Mark) and BOMs are not preferred for UTF-8 files. So, @phw and @Freso  asked me to remove it.

Comment (phw) [reply picard/ui/mainwindow.py]: Ah, sorry. That was a misunderstanding derstanding. The BOM are just three bytes at the very beginning, before the `#`. They are not visible here, this makes the discussion a bit difficult. But it go added maybe by your editor automatically. It's maybe even still in the file, despite you removed the first line.

So `# -*- coding: utf-8 -*-` should be added again. And before saving check your text editor if there is some option regarding the BOM. What editor are you using?

And to not clutter this pull request with too many commits I would suggest when doing these changes you ammend the last commit with the BOM message. That way it stays only a single commit. You have to force push the to GitHub, but for a feature branch that's ok (but never do this on a project's master branch, it will cause trouble).

Comment (abhi-ohri) [reply picard/ui/mainwindow.py]: Done.

**Comment by phw** [comment]
@abhi-ohri I think somewhat we all got confused here, sorry for that. I actually think you already got this right before your last commit and removed the BOM correctly.

But it looks like @zas got confused by the remove and add lines in e201d65f03a612c7f7932e2ffd4a12094876556e, I got confused by @zas's comment thinking it referred to your last commit and then I confused you with my comment about `#` to do the wrong thing. Sorry for that :( Stupid invisible characters.

To avoid further confusion I took the freedom to fixup and merge the branch directly. All done and merged now, thanks.
